Add device-state.py example mimicking "nmcli device"

bernhardkaindl opened the pull request:
Add an example demonstrating the use of the NetworkDeviceGeneric
and ActiveConnection classes to list the network devices manged
by NetworkManager as well as their type, state, internet
connectivitycheck state, device running the highest-priorized
default route and the identifier of the active connection profile of
that device. The output resembles the output of the NetworkManager
CLI command "nmcli device".

igo95862 commented:
Thank you for submitting more examples. It was desperately needed.

However, would you mind making changes?

igo95862 reviewed (changes requested):
The path probably should be `block` not `blok`

igo95862 commented:
Overall I am not entirely sure why the helper classes were created. Maybe my documentation is lacking.

bernhardkaindl commented:
Just for your information, while I look at this example:

I squashed the commit(s) of PR #2 now to consist of a single commit only, so you can review it. Can you take a look at it?

igo95862 commented:
Look great code wise. I will give it a test later today.

bernhardkaindl commented:
I refactored the example of this PR now completely to be based on PR #2 and your example in `docs/examples.rst`

I think this looks good and crisp now, please take a look and of course run it :-)

If this approach can find your approval, I can add an async version of this example again (it will need a number of "await"s)

igo95862 commented:
Ok. It runs great on my machine.

However, `mypy --strict` complains about untyped functions. Should be very easy fix by annotating functions with `-> None` or `-> str`

```
device-state.py:27: error: Function is missing a type annotation
device-state.py:32: error: Function is missing a return type annotation
device-state.py:32: note: Use "-> None" if function does not return a value
device-state.py:48: error: Call to untyped function "title" in typed context
device-state.py:49: error: Call to untyped function "title" in typed context
device-state.py:50: error: Call to untyped function "title" in typed context
device-state.py:63: error: Call to untyped function "list_active_hardware_networkdevice_states" in typed context
```

After that and async version I will merge the pull request. (and add it to documentation page) Thank you again for your work.

igo95862 commented:
By the way is

```
if generic_dev.capabilities & DeviceCapabilities.IS_SOFTWARE:
            continue
```

required? The `nmcli device` list software based devices (like VLANs or VPNs).

bernhardkaindl commented:
> By the way is
> ```
> if generic_dev.capabilities & DeviceCapabilities.IS_SOFTWARE:
>             continue
> ```
> required? The `nmcli device` list software based devices (like VLANs or VPNs).

@igo95862: No it's not required. In my case, they just were more interesting. I pushed an update, changes:
* List software devices as well (by default, can be selected using argparse now)
* Fixed `mypy --strict`
* Fixed the list to also show not activated devices as well (just without the `id` for the active connection of course)
* To keep the line length below 79 columns  (for flake8 and readability) for the `Capabilites.IS_SOFTWARE` check, I abbreviated `DeviceCapabilities.IS_SOFTWARE` to `Capabilities.IS_SOFTWARE` (using `import as`).

I'll do the async version once all checks and remarks for the blocking version are complete.

bernhardkaindl commented:
@igo95862 Done, the async version is pushed now as well. A quick explanation:
* For `mypy`, I had to add extra string annotations for:
```py
dev: str = await generic.interface
id: str = await active_conn.id
```
Otherwise there was nothing besides async/await I didn't forsee.
Tested and checked with `flake8` and `mypy --strict`

bernhardkaindl commented:
@igo95862: JFYI, accidentially, you have put this message into the commit message of the merge commit ;-)
"Thank you again. I will look in to other merge requests in upcoming days."

A nice option would be to use rebase for the merge, then my original commit message should be preserved.
